### ISSUE TITLE
ZMap build is requiring libnuma-dev

### ISSUE THREAD

**Issue by zakird** [zmap/zmap#447]
This issue supersedes this pull request: https://github.com/zmap/zmap/pull/446. This was originally reported by @azet. 

One some systems, ZMap incidentally tries to build with with PF_RING support. I believe that PF_RING does require libnuma, but the default should not be trying to build with PF_RING support.

I'm specifically noting:

[ 20%] Building C object src/CMakeFiles/ztests.dir/socket-pfring.c.o
[ 20%] Building C object src/CMakeFiles/ztests.dir/recv-pfring.c.o
But if you look at the CMake configuration, you'll see

# Handle various versions of socket
if(WITH_PFRING)
    set(SOURCES ${SOURCES} socket-pfring.c)
    set(ZTESTSOURCES ${ZTESTSOURCES} socket-pfring.c)
elseif (APPLE OR BSD)
    set(SOURCES ${SOURCES} socket-bsd.c)
    set(ZTESTSOURCES ${ZTESTSOURCES} socket-bsd.c)
else()
    set(SOURCES ${SOURCES} socket-linux.c)
    set(ZTESTSOURCES ${ZTESTSOURCES} socket-linux.c)
endif()
So I think this issue should really be: Why is @azet's system think WITH_PFRING is set when it's not.

@azet just to confirm that this isn't something leftover, can you try to clone to a fresh directory and rebuild and see if you're trying to build with PF_RING support?

**Comment by azet** [comment]
> ZMap incidentally tries to build with with PF_RING support

This wasn't what I was reporting! I was building with `PF_RING` ZeroCache support on purpose. I don't think ZMap will build without the special `cmake` flags set. I may be missing something?

**Comment by zakird** [comment]
OK. Got it. Unfortunately, this is contrary to what you posted in the issue :) : 

```
azet@scan-sha2017:~/software/zmap/v2.1.1$ cmake .
-- Checking for module 'json-c'
--   Found json-c, version 0.12.1
-- Default ZMap configuration file location is /etc/zmap
-- Checking if blacklist.conf exists there...
-- Checking if zmap.conf exists there...
-- Configuring done
-- Generating done
-- Build files have been written to: /home/azet/software/zmap/v2.1.1
```

If this is indeed the problem, then we just need to add this as a documented dependency for 10gigE scanning at: https://github.com/zmap/zmap/blob/master/10gigE.md

**Comment by azet** [comment]
@zakird 

Verified:

```
azet@scan-sha2017:~/software$ git clone https://github.com/zmap/zmap zmap-2 
Cloning into 'zmap-2'...
remote: Counting objects: 5880, done.
remote: Compressing objects: 100% (53/53), done.
remote: Total 5880 (delta 48), reused 15 (delta 8), pack-reused 5819
Receiving objects: 100% (5880/5880), 5.53 MiB | 5.58 MiB/s, done.
Resolving deltas: 100% (4135/4135), done.
azet@scan-sha2017:~/software$ cd zmap-2/
azet@scan-sha2017:~/software/zmap-2$ cmake .
-- The C compiler identification is GNU 6.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29") 
-- Checking for module 'json-c'
--   Found json-c, version 0.12.1
-- Default ZMap configuration file location is /etc/zmap
-- Checking if blacklist.conf exists there...
CMake Warning at CMakeLists.txt:156 (message):
  Existing configuration file detected at /etc/zmap/blacklist.conf,
  blacklist.conf from sources will NOT be installed.  Please check and
  install manually!


-- Checking if zmap.conf exists there...
CMake Warning at CMakeLists.txt:156 (message):
  Existing configuration file detected at /etc/zmap/zmap.conf, zmap.conf from
  sources will NOT be installed.  Please check and install manually!


-- Configuring done
-- Generating done
-- Build files have been written to: /home/azet/software/zmap-2
azet@scan-sha2017:~/software/zmap-2$ make -j4
Scanning dependencies of target git_versioning
Scanning dependencies of target zmaplib
[  1%] Building C object lib/CMakeFiles/zmaplib.dir/blacklist.c.o
[  1%] Building C object lib/CMakeFiles/zmaplib.dir/constraint.c.o
[  1%] Built target git_versioning
[  2%] Building C object lib/CMakeFiles/zmaplib.dir/logger.c.o
[  3%] Building C object lib/CMakeFiles/zmaplib.dir/pbm.c.o
[  4%] Building C object lib/CMakeFiles/zmaplib.dir/rijndael-alg-fst.c.o
[  5%] Building C object lib/CMakeFiles/zmaplib.dir/random.c.o
[  6%] Building C object lib/CMakeFiles/zmaplib.dir/xalloc.c.o
[  7%] Building C object lib/CMakeFiles/zmaplib.dir/lockfd.c.o
[  8%] Building C object lib/CMakeFiles/zmaplib.dir/util.c.o
[  9%] Building C object lib/CMakeFiles/zmaplib.dir/queue.c.o
[ 10%] Building C object lib/CMakeFiles/zmaplib.dir/csv.c.o
[ 11%] Linking C static library libzmaplib.a
[ 11%] Built target zmaplib
[ 11%] Generating zbopt.h
[ 12%] Generating parser.c
[ 13%] Generating parser.c
[ 14%] Generating zitopt.h
[ 16%] Generating zopt.h
[ 16%] Generating ztopt.h
Scanning dependencies of target zblacklist
[ 17%] Generating lexer.c
Scanning dependencies of target ziterate
[ 18%] Generating lexer.c
[ 19%] Building C object src/CMakeFiles/zblacklist.dir/zblacklist.c.o
[ 20%] Building C object src/CMakeFiles/ziterate.dir/aesrand.c.o
Scanning dependencies of target zmap
[ 21%] Building C object src/CMakeFiles/ziterate.dir/cyclic.c.o
Scanning dependencies of target ztests
[ 22%] Building C object src/CMakeFiles/zmap.dir/aesrand.c.o
[ 24%] Building C object src/CMakeFiles/ztests.dir/aesrand.c.o
[ 24%] Building C object src/CMakeFiles/zblacklist.dir/zbopt_compat.c.o
In file included from /home/azet/software/zmap-2/src/zbopt_compat.c:15:0:
/home/azet/software/zmap-2/src/zbopt.c: In function 'cmdline_parser_internal':
/home/azet/software/zmap-2/src/zbopt.c:505:7: warning: variable 'check_required' set but not used [-Wunused-but-set-variable]
   int check_required;
       ^~~~~~~~~~~~~~
[ 25%] Building C object src/CMakeFiles/zmap.dir/cyclic.c.o
[ 26%] Building C object src/CMakeFiles/ztests.dir/cyclic.c.o
[ 27%] Building C object src/CMakeFiles/ziterate.dir/iterator.c.o
[ 28%] Building C object src/CMakeFiles/zmap.dir/expression.c.o
[ 29%] Building C object src/CMakeFiles/ztests.dir/expression.c.o
[ 30%] Linking C executable zblacklist
[ 31%] Building C object src/CMakeFiles/ziterate.dir/shard.c.o
[ 32%] Building C object src/CMakeFiles/ztests.dir/fieldset.c.o
[ 33%] Building C object src/CMakeFiles/zmap.dir/fieldset.c.o
[ 33%] Built target zblacklist
[ 34%] Building C object src/CMakeFiles/zmap.dir/filter.c.o
[ 35%] Building C object src/CMakeFiles/ziterate.dir/state.c.o
[ 36%] Building C object src/CMakeFiles/ziterate.dir/validate.c.o
[ 37%] Building C object src/CMakeFiles/zmap.dir/get_gateway.c.o
[ 38%] Building C object src/CMakeFiles/ziterate.dir/zitopt_compat.c.o
In file included from /home/azet/software/zmap-2/src/zitopt_compat.c:15:0:
/home/azet/software/zmap-2/src/zitopt.c: In function 'cmdline_parser_internal':
/home/azet/software/zmap-2/src/zitopt.c:545:7: warning: variable 'check_require' set but not used [-Wunused-but-set-variable]
   int check_required;
       ^~~~~~~~~~~~~~
[ 39%] Building C object src/CMakeFiles/ziterate.dir/ziterate.c.o
[ 40%] Building C object src/CMakeFiles/ztests.dir/filter.c.o
[ 41%] Building C object src/CMakeFiles/ztests.dir/get_gateway.c.o
[ 42%] Building C object src/CMakeFiles/ztests.dir/iterator.c.o
[ 43%] Building C object src/CMakeFiles/zmap.dir/iterator.c.o
[ 43%] Linking C executable ziterate
[ 44%] Building C object src/CMakeFiles/zmap.dir/monitor.c.o
[ 45%] Generating topt.h
[ 46%] Building C object src/CMakeFiles/ztests.dir/monitor.c.o
Scanning dependencies of target ztee
[ 46%] Built target ziterate
[ 47%] Building C object src/CMakeFiles/ztee.dir/ztee.c.o
[ 48%] Building C object src/CMakeFiles/zmap.dir/recv.c.o
[ 48%] Building C object src/CMakeFiles/ztee.dir/topt_compat.c.o
[ 49%] Building C object src/CMakeFiles/zmap.dir/send.c.o
In file included from /home/azet/software/zmap-2/src/topt_compat.c:7:0:
/home/azet/software/zmap-2/src/topt.c: In function 'update_arg':
/home/azet/software/zmap-2/src/topt.c:397:9: warning: variable 'stop_char' set but not used [-Wunused-but-set-variable]
   char *stop_char = 0;
         ^~~~~~~~~
/home/azet/software/zmap-2/src/topt.c: In function 'cmdline_parser_internal':
/home/azet/software/zmap-2/src/topt.c:476:7: warning: variable 'check_required' set but not used [-Wunused-but-set-variable]
   int check_required;
       ^~~~~~~~~~~~~~
[ 50%] Building C object src/CMakeFiles/ztests.dir/recv.c.o
[ 50%] Building C object src/CMakeFiles/zmap.dir/shard.c.o
[ 50%] Building C object src/CMakeFiles/ztests.dir/send.c.o
[ 51%] Building C object src/CMakeFiles/ztests.dir/shard.c.o
[ 52%] Linking C executable ztee
[ 53%] Building C object src/CMakeFiles/zmap.dir/socket.c.o
[ 53%] Built target ztee
[ 54%] Building C object src/CMakeFiles/zmap.dir/state.c.o
[ 55%] Building C object src/CMakeFiles/zmap.dir/summary.c.o
[ 56%] Building C object src/CMakeFiles/ztests.dir/socket.c.o
[ 57%] Building C object src/CMakeFiles/zmap.dir/validate.c.o
[ 58%] Building C object src/CMakeFiles/zmap.dir/zmap.c.o
[ 59%] Building C object src/CMakeFiles/ztests.dir/state.c.o
[ 60%] Building C object src/CMakeFiles/ztests.dir/summary.c.o
[ 61%] Building C object src/CMakeFiles/ztests.dir/validate.c.o
[ 62%] Building C object src/CMakeFiles/ztests.dir/ztopt_compat.c.o
[ 63%] Building C object src/CMakeFiles/ztests.dir/probe_modules/module_icmp_echo.c.o
In file included from /home/azet/software/zmap-2/src/ztopt_compat.c:15:0:
/home/azet/software/zmap-2/src/ztopt.c: In function 'update_arg':
/home/azet/software/zmap-2/src/ztopt.c:345:15: warning: variable 'val' set but not used [-Wunused-but-set-variable]
   const char *val = value;
               ^~~
/home/azet/software/zmap-2/src/ztopt.c:344:9: warning: variable 'stop_char' set but not used [-Wunused-but-set-variable]
   char *stop_char = 0;
         ^~~~~~~~~
/home/azet/software/zmap-2/src/ztopt.c: In function 'cmdline_parser_internal':
/home/azet/software/zmap-2/src/ztopt.c:414:7: warning: variable 'check_required' set but not used [-Wunused-but-set-variable]
   int check_required;
       ^~~~~~~~~~~~~~
[ 64%] Building C object src/CMakeFiles/zmap.dir/zopt_compat.c.o
In file included from /home/azet/software/zmap-2/src/zopt_compat.c:15:0:
/home/azet/software/zmap-2/src/zopt.c: In function 'cmdline_parser_internal':
/home/azet/software/zmap-2/src/zopt.c:854:7: warning: variable 'check_required' set but not used [-Wunused-but-set-variable]
   int check_required;
       ^~~~~~~~~~~~~~
[ 65%] Building C object src/CMakeFiles/zmap.dir/lexer.c.o
/home/azet/software/zmap-2/src/lexer.c: In function 'yy_get_next_buffer':
/home/azet/software/zmap-2/src/lexer.c:586:18: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
   for ( n = 0; n < max_size && \
                  ^
/home/azet/software/zmap-2/src/lexer.c:1060:3: note: in expansion of macro 'YY_INPUT'
   YY_INPUT( (&YY_CURRENT_BUFFER_LVALUE->yy_ch_buf[number_to_move]),
   ^~~~~~~~
[ 66%] Building C object src/CMakeFiles/zmap.dir/parser.c.o
[ 67%] Building C object src/CMakeFiles/ztests.dir/probe_modules/module_icmp_echo_time.c.o
[ 68%] Building C object src/CMakeFiles/ztests.dir/probe_modules/module_tcp_synscan.c.o
[ 69%] Building C object src/CMakeFiles/zmap.dir/probe_modules/module_icmp_echo.c.o
[ 70%] Building C object src/CMakeFiles/ztests.dir/probe_modules/module_udp.c.o
[ 71%] Building C object src/CMakeFiles/ztests.dir/probe_modules/module_tcp_synackscan.c.o
[ 72%] Building C object src/CMakeFiles/ztests.dir/probe_modules/packet.c.o
[ 73%] Building C object src/CMakeFiles/zmap.dir/probe_modules/module_icmp_echo_time.c.o
[ 74%] Building C object src/CMakeFiles/zmap.dir/probe_modules/module_tcp_synscan.c.o
[ 75%] Building C object src/CMakeFiles/zmap.dir/probe_modules/module_tcp_synackscan.c.o
[ 75%] Building C object src/CMakeFiles/ztests.dir/probe_modules/probe_modules.c.o
[ 75%] Building C object src/CMakeFiles/zmap.dir/probe_modules/module_udp.c.o
[ 76%] Building C object src/CMakeFiles/zmap.dir/probe_modules/packet.c.o
[ 77%] Building C object src/CMakeFiles/ztests.dir/probe_modules/module_ntp.c.o
[ 78%] Building C object src/CMakeFiles/zmap.dir/probe_modules/probe_modules.c.o
[ 79%] Building C object src/CMakeFiles/zmap.dir/probe_modules/module_ntp.c.o
[ 80%] Building C object src/CMakeFiles/zmap.dir/probe_modules/module_upnp.c.o
[ 81%] Building C object src/CMakeFiles/ztests.dir/probe_modules/module_upnp.c.o
[ 82%] Building C object src/CMakeFiles/zmap.dir/probe_modules/module_dns.c.o
[ 83%] Building C object src/CMakeFiles/ztests.dir/probe_modules/module_dns.c.o
[ 84%] Building C object src/CMakeFiles/zmap.dir/probe_modules/module_bacnet.c.o
[ 85%] Building C object src/CMakeFiles/ztests.dir/probe_modules/module_bacnet.c.o
[ 86%] Building C object src/CMakeFiles/ztests.dir/output_modules/module_csv.c.o
[ 87%] Building C object src/CMakeFiles/ztests.dir/output_modules/module_json.c.o
/home/azet/software/zmap-2/src/probe_modules/module_dns.c: In function 'dns_global_initialize':
/home/azet/software/zmap-2/src/probe_modules/module_dns.c:527:12: warning: stack protector not protecting local variables: variable length buffer [-Wstack-protector]
 static int dns_global_initialize(struct state_conf *conf)
            ^~~~~~~~~~~~~~~~~~~~~
/home/azet/software/zmap-2/src/probe_modules/module_dns.c: In function 'dns_global_initialize':
/home/azet/software/zmap-2/src/probe_modules/module_dns.c:527:12: warning: stack protector not protecting local variables: variable length buffer [-Wstack-protector]
 static int dns_global_initialize(struct state_conf *conf)
            ^~~~~~~~~~~~~~~~~~~~~
[ 88%] Building C object src/CMakeFiles/zmap.dir/output_modules/module_csv.c.o
[ 89%] Building C object src/CMakeFiles/zmap.dir/output_modules/module_json.c.o
[ 90%] Building C object src/CMakeFiles/ztests.dir/output_modules/output_modules.c.o
[ 91%] Building C object src/CMakeFiles/ztests.dir/tests/test_harness.c.o
[ 92%] Building C object src/CMakeFiles/zmap.dir/socket-linux.c.o
[ 93%] Building C object src/CMakeFiles/zmap.dir/output_modules/output_modules.c.o
[ 94%] Building C object src/CMakeFiles/zmap.dir/recv-pcap.c.o
[ 95%] Building C object src/CMakeFiles/ztests.dir/lexer.c.o
[ 96%] Building C object src/CMakeFiles/ztests.dir/socket-linux.c.o
/home/azet/software/zmap-2/src/lexer.c: In function 'yy_get_next_buffer':
/home/azet/software/zmap-2/src/lexer.c:586:18: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
   for ( n = 0; n < max_size && \
                  ^
/home/azet/software/zmap-2/src/lexer.c:1060:3: note: in expansion of macro 'YY_INPUT'
   YY_INPUT( (&YY_CURRENT_BUFFER_LVALUE->yy_ch_buf[number_to_move]),
   ^~~~~~~~
[ 97%] Building C object src/CMakeFiles/ztests.dir/parser.c.o
[ 98%] Building C object src/CMakeFiles/ztests.dir/recv-pcap.c.o
[ 99%] Linking C executable zmap
[ 99%] Built target zmap
[100%] Linking C executable ztests
[100%] Built target ztests
azet@scan-sha2017:~/software/zmap-2$ dpkg -l libnuma-dev
dpkg-query: no packages found matching libnuma-dev
azet@scan-sha2017:~/software/zmap-2$ 

```

**Comment by azet** [comment]
> Unfortunately, this is contrary to what you posted in the issue :) :

Well, yes and no (it was very late already). So I was trying to re-build **without** 10GE support afterwards, but running `cmake` like for a usual installation was using the previous options (I think because of the cmake cache, and no proper `cmake clean` command).

**Comment by zakird** [comment]
If you want to start over, your best bet is going to be something like running `git clean -dfx` to take care of old CMake files. CMake doesn't have a great clean option, but this would have been the source of your problems.

**Comment by zakird** [comment]
This drives me nuts about cmake. It might be good idea to include a shell script in the repo that does the right thing given that running `git clean` isn't very reasoanble.

**Comment by azet** [comment]
@zakird 
> If you want to start over, your best bet is going to be something like running git clean -dfx to take care of old CMake files. CMake doesn't have a great clean option, but this would have been the source of your problems.

Yea, this is what I did in the end. Well, `rm -rf zmap/ && git clone ..` is what I did. :)

Why are you depending on `cmake` anyway? I don't see an urgent reason to use it for the build process here. In my experience`cmake` causes more troubles than it solves (especially when used with [poor or old] scientific software, your project - of course - excluded)..

**Comment by zakird** [comment]
OS distribution. We don't love it, but it seems to make the folks who distribute ZMap much happier, and we want to make their lives as easy as possible.

**Comment by azet** [comment]
@zakird 

> This drives me nuts about cmake. It might be good idea to include a shell script in the repo that does the right thing given that running git clean isn't very reasoanble.
>  

Many projects ship a simple shell or bash script called `bootstrap` that takes care of initial bootstrapping, checking for missing dependencies et cetera. Maybe that's an option for developers trying to build from source?

**Comment by zakird** [comment]
Realistically, almost everyone should be installing this through a package manager not building from source. We don't particularly want to be encouraging folks to be installing from source unless they're developers, in which case, having strong and correct documentation seems appropriate.

**Comment by azet** [comment]
@zakird 

Sure, and that's what your README / Install documents say. I totally agree that distro packages are the best way to go. But some people will need to build from source (e.g. for fixing bugs, improving parts of the code or benchmarking/testing as in my case).

**Comment by dadrian** [comment]
I added documentation to 10GigE in commit 624e19d.

**Comment by zakird** [comment]
@azet Yep, absolutely agree.

Out of curiosity, were you able to get 10gigE scanning working? Given the the large packet drop at destination networks seen when performing seen doing scans >3gbps, we've somewhat neglected the 10gigE scanning portion of ZMap the last few years.

**Comment by azet** [comment]
We'll I'm dependent on hardware from other people and a quite spontaneous (i.e. non-datacenter) networking / NOC at sha2017.org - the server isn't mine and I had little time for network tuning nor any influence on the 10GE swithes, BGP info etc. 

I'm currently having an issue with the specific Intel card we've borrowed and `PF_RING` zerocopy kernel modules, in addition, the server only has 8GB of RAM (for ~20 scans in parallel), a single 8core Xeon E5v3 @ 3.2GHZ and no out of band management. I'm doing my best, but this isn't what I'd call a proper lab setup nor would I usually benchmark in such an environment. But we've got a lot of free upstream bandwidth, so why not grab as much as we can as best as we can improvise on site? People came up with cool ideas here out of the blue. I'm also scanning iWARP, RoCEv2 and SCTP for the first time.

see:
- https://gist.github.com/azet/54862407b7af1c2813e590ead83f7553
- https://wiki.sha2017.org/index.php?title=Internet-wide_Scanning
- scan-sha2017.external.meta.southbound.io:8080 (public:guest)

If you've got any ideas and want to try them out send me a ssh public key to `shascan@azet.org` and a submission proposal (which we'll accept right away from you guys anyways, it's intended for people with little to no experience in internet measurement). Attribution and data-sets go to the original authors.

We've also a few custom scans I cannot tell much about at the moment that we'd like to re-run or test.
